Make iframes block the enclosing document's load event

jdm opened the pull request:
It occurs to me as I write this that this doesn't handle the case of removing the iframe from the document before it's finished loading. Consider this an early feedback release!

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/6677)

<!-- Reviewable:end -->

jdm committed:
Make iframes block the enclosing document's load event. Fixes #6663.

Ms2ger commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit f812e7f has been approved by `Ms2ger`

<!-- @bors-servo r=Ms2ger f812e7fbc6b4e62594a20cf2bccf7d31639fa657 -->

bors-servo commented:
:hourglass: Testing commit f812e7f with merge 0fdb2aa...

bors-servo commented:
:broken_heart: Test failed - [mac2](http://build.servo.org/builders/mac2/builds/2570)

Ms2ger commented:
```
failures:
    iframe/hide_layers1.html == iframe/hide_layers_ref.html
```

bors-servo commented:
:umbrella: The latest upstream changes (presumably #6682) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Interestingly, the test fails because while the iframe borders disappear, the iframe content remains. This is unaffected by the presence of `-i`.

glennw commented:
@jdm This means that the owning document has correctly hidden the iframe node, but that the compositor is still drawing the iframe layer, if that's of any help.

metajack commented:
r? @Ms2ger

jdm commented:
I think I'd like to finish the missing bits of unblocking the page if the iframe is removed before merging this again.

jdm committed:
Add test for unblocking onload when removing an iframe from the tree.

jdm commented:
@Ms2ger I've made all the necessary changes; this is ready for review again.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7591) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8088) made this pull request unmergeable. Please resolve the merge conflicts.

Ms2ger commented:
Reviewed 2 of 52 files at r1, 1 of 9 files at r2, 1 of 2 files at r3.
Review status: 2 of 53 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/document_loader.rs, line 40 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6677#-K15-JWEudStFIVs80al)** ([raw file](https://github.com/servo/servo/blob/e4cf4c64118d0fc08747448dacd4c4dca5351c6c/components/script/document_loader.rs#L40)):</sup>
"Helpful" is not a useful word here; remove it? "destructor bomb" is not exactly standard nomenclature; try to find a better name, please.
<!-- Reviewable comment -K15-JWEudStFIVs80am:-1433319126 -->

---

<sup>**[components/script/document_loader.rs, line 52 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6677#-K15-eeSdH33LpDMQEjr)** ([raw file](https://github.com/servo/servo/blob/e4cf4c64118d0fc08747448dacd4c4dca5351c6c/components/script/document_loader.rs#L52)):</sup>
Docs for the methods
<!-- Reviewable comment -K15-eeTQL3Vy-9CvcsC:1151833836 -->

---

<sup>**[components/script/document_loader.rs, line 61 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6677#-K15-EQqUyVy0BYpJlrS)** ([raw file](https://github.com/servo/servo/blob/e4cf4c64118d0fc08747448dacd4c4dca5351c6c/components/script/document_loader.rs#L61)):</sup>
No need; you `unwrap()` below.
<!-- Reviewable comment -K15-EQrZUiLvYkazPNB:-362168277 -->

---

<sup>**[components/script/dom/document.rs, line 980 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6677#-K14z-iHDJfFPQtHIHzx)** ([raw file](https://github.com/servo/servo/blob/5dea7c0c31fe1aa7791aa3c39a6f182bef89989c/components/script/dom/document.rs#L980)):</sup>
Document which steps of this algorithm this method covers.
<!-- Reviewable comment -K14z-iImtHv0V-f0RF8:774086319 -->

---

<sup>**[components/script/dom/document.rs, line 986 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6677#-K14xmfbUkAU2Xc5irmq)** ([raw file](https://github.com/servo/servo/blob/5dea7c0c31fe1aa7791aa3c39a6f182bef89989c/components/script/dom/document.rs#L986)):</sup>
Reuse `find_iframe()`.
<!-- Reviewable comment -K14xmfbUkAU2Xc5irmr:-1875694934 -->

---

<sup>**[components/script/dom/document.rs, line 1940 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6677#-K14x_2THZp9asvU8_cc)** ([raw file](https://github.com/servo/servo/blob/5dea7c0c31fe1aa7791aa3c39a6f182bef89989c/components/script/dom/document.rs#L1940)):</sup>
We already send `DOMLoad`, why don't we piggy-back on that?
<!-- Reviewable comment -K14x_2THZp9asvU8_cd:-442417790 -->

---

<sup>**[components/script/dom/htmliframeelement.rs, line 222 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6677#-K15-omzfFzo1o8FCUP6)** ([raw file](https://github.com/servo/servo/blob/e4cf4c64118d0fc08747448dacd4c4dca5351c6c/components/script/dom/htmliframeelement.rs#L222)):</sup>
Again, steps.
<!-- Reviewable comment -K15-omzfFzo1o8FCUP7:-686379338 -->

---

<sup>**[components/script/dom/htmliframeelement.rs, line 239 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6677#-K1504Vp188fRP71KCbE)** ([raw file](https://github.com/servo/servo/blob/e4cf4c64118d0fc08747448dacd4c4dca5351c6c/components/script/dom/htmliframeelement.rs#L239)):</sup>
Consider `take()` here and below.
<!-- Reviewable comment -K1504Vp188fRP71KCbF:1995658192 -->

---

<sup>**[components/script/dom/htmliframeelement.rs, line 453 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6677#-K15-uL4kPRufHkqO1Sr)** ([raw file](https://github.com/servo/servo/blob/e4cf4c64118d0fc08747448dacd4c4dca5351c6c/components/script/dom/htmliframeelement.rs#L453)):</sup>
You should be able to drop the two `&mut`s now.
<!-- Reviewable comment -K15-uL4kPRufHkqO1Ss:-596325442 -->

---

<sup>**[tests/wpt/mozilla/tests/mozilla/iframe-unblock-onload.html, line 18 \[r3\]](https://reviewable.io:443/reviews/servo/servo/6677#-K150YlH_JckBqzeqDlg)** ([raw file](https://github.com/servo/servo/blob/13161e160de6802d08f25710a9db247ad458eb8a/tests/wpt/mozilla/tests/mozilla/iframe-unblock-onload.html#L18)):</sup>
Just remove the manual timeout; having the failure mode be `TIMEOUT` is perfectly fine.
<!-- Reviewable comment -K150YlH_JckBqzeqDlh:-1986686159 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6677)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8115) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Rebased and addressed.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 238d194 with merge 999168d...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/634)

jdm commented:
Clearly needs some more work.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8144) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit e07bfb7 with merge 5097372...

larsbergstrom commented:
@bors-servo retry

Manishearth commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit e07bfb7 with merge 275f8ef...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1)

jdm commented:
@bors-servo: try

jdm commented:
@bors-servo: retry

jdm commented:
@bors-servo: try

bors-servo commented:
:lock: Merge conflict

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 1bd9364 with merge ebdff2c...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/43)

jdm commented:
From linux:
```
/XMLHttpRequest/send-entity-body-document.htm
---------------------------------------------
FAIL HTML document, invalid UTF-8
FAIL plain text file
FAIL img tag
FAIL empty div
FAIL HTML document, shift-jis
FAIL image file
/dom/nodes/Document-URL.sub.html
--------------------------------
TIMEOUT [Parent]
/dom/nodes/Document-createElement-namespace.html
------------------------------------------------
TIMEOUT [Parent]
/dom/nodes/Node-parentNode.html
-------------------------------
TIMEOUT [Parent]
/html/browsers/browsing-the-web/navigating-across-documents/013.html
--------------------------------------------------------------------
CRASH [Parent]
/html/browsers/browsing-the-web/navigating-across-documents/015.html
--------------------------------------------------------------------
CRASH [Parent]
/html/browsers/history/the-location-interface/location_reload.html
------------------------------------------------------------------
CRASH [Parent]
/2dcontext/drawing-images-to-the-canvas/drawimage_canvas_12.html
----------------------------------------------------------------
TIMEOUT [Parent]
```

jdm commented:
From mac, iframe/hide_layers2.html failed in the reftests.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8094) made this pull request unmergeable. Please resolve the merge conflicts.

eefriedman commented:
(The failures on `/XMLHttpRequest/send-entity-body-document.htm` are expected; see also #8157.)

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit a2b63ee with merge db2087a...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/1035)

jdm commented:
@bors-servo: retry

bors-servo commented:
:hourglass: Trying commit a2b63ee with merge 0aacf84...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1083)

jdm committed:
Reorder marking of pending parser-blocking scripts to avoid reentrancy.

In cases where a pending parser-blocking script is executing, if the script
caused another load to complete (eg. by removing a pending iframe from the
DOM, thus terminating the load) then the currently-executing script would
attempt to execute itself again.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit bc79405 with merge b00473e...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/1171)

jdm commented:
```
  ▶ Unexpected subtest result in /XMLHttpRequest/send-entity-body-document.htm:
  │ FAIL [expected PASS] HTML document, shift-jis
  │   → assert_equals: document should be serialized and sent as text/html;charset=UTF-8 (TEST#2) expected "text/html;charset=UTF-8" but got "text/plain;charset=UTF-8"
  │ FAIL [expected PASS] plain text file
  │   → assert_equals: document should be serialized and sent as text/html;charset=UTF-8 (TEST#3) expected "text/html;charset=UTF-8" but got "text/plain;charset=UTF-8"
  │ FAIL [expected PASS] empty div
  │   → assert_equals: document should be serialized and sent as text/html;charset=UTF-8 (TEST#6) expected "text/html;charset=UTF-8" but got "text/plain;charset=UTF-8"
  │ FAIL [expected PASS] img tag
  │   → assert_equals: document should be serialized and sent as text/html;charset=UTF-8 (TEST#5) expected "text/html;charset=UTF-8" but got "text/plain;charset=UTF-8"
  │ FAIL [expected PASS] image file
  │   → assert_equals: document should be serialized and sent as text/html;charset=UTF-8 (TEST#4) expected "text/html;charset=UTF-8" but got "text/plain;charset=UTF-8"
  │ 
  │ request/<@http://web-platform.test:8000/XMLHttpRequest/send-entity-body-document.htm:18:1
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1381:20
  │ test@http://web-platform.test:8000/resources/testharness.js:495:9
  │ request@http://web-platform.test:8000/XMLHttpRequest/send-entity-body-document.htm:13:1
  │ init@http://web-platform.test:8000/XMLHttpRequest/send-entity-body-document.htm:29:42
  └ load@http://web-platform.test:8000/XMLHttpRequest/send-entity-body-document.htm:0:1

  ▶ Unexpected subtest result in /dom/ranges/Range-deleteContents.html:
  │ FAIL [expected PASS] Resulting cursor position for range 18 [paras[0].firstChild, 0, paras[1].firstChild, 0]
  │   → assert_equals: Unexpected startOffset after deleteContents() expected 1 but got 0
  │ FAIL [expected PASS] Resulting cursor position for range 19 [paras[0].firstChild, 0, paras[1].firstChild, 8]
  │   → assert_equals: Unexpected startOffset after deleteContents() expected 1 but got 0
  │ FAIL [expected PASS] Resulting cursor position for range 20 [paras[0].firstChild, 3, paras[3], 1]
  │   → assert_equals: Unexpected startOffset after deleteContents() expected 1 but got 0
  │ FAIL [expected PASS] Resulting cursor position for range 50 [paras[2].firstChild, 4, comment, 2]
  │   → assert_equals: Unexpected startOffset after deleteContents() expected 3 but got 2
  │ FAIL [expected PASS] Resulting cursor position for range 51 [paras[3], 1, comment, 8]
  │   → assert_equals: Unexpected startOffset after deleteContents() expected 4 but got 3
  │ 
  │ testDeleteContents/<@http://web-platform.test:8000/dom/ranges/Range-deleteContents.html:259:1
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1381:20
  │ testDeleteContents@http://web-platform.test:8000/dom/ranges/Range-deleteContents.html:238:3
  └ actualIframe.onload/expectedIframe.onload@http://web-platform.test:8000/dom/ranges/Range-deleteContents.html:318:7

  ▶ Unexpected subtest result in /dom/ranges/Range-deleteContents.html:
  │ FAIL [expected PASS] Resulting DOM for range 10 [document.documentElement, 0, document.documentElement, 1]
  │   → assert_equals: Actual and expected DOMs were broken up into a different number of pieces by deleteContents() (this probably means you created or detached nodes when you weren't supposed to) expected 2 but got 3
  │ FAIL [expected PASS] Resulting DOM for range 11 [document.documentElement, 0, document.documentElement, 2]
  │   → assert_equals: Actual and expected DOMs were broken up into a different number of pieces by deleteContents() (this probably means you created or detached nodes when you weren't supposed to) expected 3 but got 12
  │ FAIL [expected PASS] Resulting DOM for range 12 [document.documentElement, 1, document.documentElement, 2]
  │   → assert_equals: Actual and expected DOMs were broken up into a different number of pieces by deleteContents() (this probably means you created or detached nodes when you weren't supposed to) expected 2 but got 10
  │ FAIL [expected PASS] Resulting DOM for range 15 [foreignDoc.documentElement, 0, foreignDoc.documentElement, 1]
  │   → assert_equals: Actual and expected DOMs were broken up into a different number of pieces by deleteContents() (this probably means you created or detached nodes when you weren't supposed to) expected 2 but got 3
  │ FAIL [expected PASS] Resulting DOM for range 24 [document, 0, document, 2]
  │   → assert_equals: Actual and expected DOMs were broken up into a different number of pieces by deleteContents() (this probably means you created or detached nodes when you weren't supposed to) expected 3 but got 20
  │ FAIL [expected PASS] Resulting DOM for range 27 [foreignDoc, 1, foreignComment, 2]
  │   → assert_equals: Actual and expected DOMs were broken up into a different number of pieces by deleteContents() (this probably means you created or detached nodes when you weren't supposed to) expected 2 but got 6
  │ FAIL [expected PASS] Resulting DOM for range 49 [document, 1, document, 2]
  │   → assert_equals: Actual and expected DOMs were broken up into a different number of pieces by deleteContents() (this probably means you created or detached nodes when you weren't supposed to) expected 2 but got 19
  │ 
  │ testDeleteContents/<@http://web-platform.test:8000/dom/ranges/Range-deleteContents.html:231:1
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1381:20
  │ testDeleteContents@http://web-platform.test:8000/dom/ranges/Range-deleteContents.html:151:3
  └ actualIframe.onload/expectedIframe.onload@http://web-platform.test:8000/dom/ranges/Range-deleteContents.html:318:7

  ▶ Unexpected subtest result in /dom/ranges/Range-extractContents.html:
  │ FAIL [expected PASS] Resulting cursor position for range 18 [paras[0].firstChild, 0, paras[1].firstChild, 0]
  │   → assert_true: Unexpected startContainer after extractContents(), expected div but got body expected true got false
  │ FAIL [expected PASS] Resulting cursor position for range 19 [paras[0].firstChild, 0, paras[1].firstChild, 8]
  │   → assert_true: Unexpected startContainer after extractContents(), expected div but got body expected true got false
  │ FAIL [expected PASS] Resulting cursor position for range 20 [paras[0].firstChild, 3, paras[3], 1]
  │   → assert_true: Unexpected startContainer after extractContents(), expected div but got body expected true got false
  │ 
  │ testExtractContents/<@http://web-platform.test:8000/dom/ranges/Range-extractContents.html:155:1
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1381:20
  │ testExtractContents@http://web-platform.test:8000/dom/ranges/Range-extractContents.html:122:3
  └ actualIframe.onload/expectedIframe.onload@http://web-platform.test:8000/dom/ranges/Range-extractContents.html:229:7

  ▶ Unexpected subtest result in /dom/ranges/Range-extractContents.html:
  │ FAIL [expected PASS] Resulting cursor position for range 50 [paras[2].firstChild, 4, comment, 2]
  │   → assert_equals: Unexpected startOffset after extractContents() expected 3 but got 1
  │ FAIL [expected PASS] Resulting cursor position for range 51 [paras[3], 1, comment, 8]
  │   → assert_equals: Unexpected startOffset after extractContents() expected 4 but got 1
  │ 
  │ testExtractContents/<@http://web-platform.test:8000/dom/ranges/Range-extractContents.html:149:1
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1381:20
  │ testExtractContents@http://web-platform.test:8000/dom/ranges/Range-extractContents.html:122:3
  └ actualIframe.onload/expectedIframe.onload@http://web-platform.test:8000/dom/ranges/Range-extractContents.html:229:7

  ▶ CRASH [expected OK] /html/browsers/history/the-location-interface/location_reload.html
  │ 
  │ thread 'PaintWorker' panicked at 'assertion failed: thread::panicking() || result == 0', /home/servo/.cargo/git/checkouts/ipc-channel-3840821e54ea11a6/master/platform/linux/mod.rs:103
  │ stack backtrace:
  │ thread 'PaintWorker' panicked at 'assertion failed: thread::panicking() || result == 0', /home/servo/.cargo/git/checkouts/ipc-channel-3840821e54ea11a6/master/platform/linux/mod.rs:103
  │    1:     0x7f8b5224ad40 - <unknown>
  │    2:     0x7f8b5224dba5 - <unknown>
  │    3:     0x7f8b5224d620 - <unknown>
  │    4:     0x7f8b52237013 - <unknown>
  │    5:     0x7f8b5220d40f - <unknown>
  │    6:     0x7f8b52206cc0 - <unknown>
  │    7:     0x7f8b514c60ac - <unknown>
  │    8:     0x7f8b5224a1e8 - <unknown>
  │    9:     0x7f8b522465bb - <unknown>
  │   10:     0x7f8b514c6966 - <unknown>
  │   11:     0x7f8b5224c503 - <unknown>
  │   12:     0x7f8b4e12c181 - <unknown>
  │   13:     0x7f8b5034347c - <unknown>
  │   14:                0x0 - <unknown>

```

nox commented:
I need to look into these range failures.

nox commented:
Ah no, that's just a badly-rebased PR.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit fcffcf0 with merge c22f0fd...

jdm commented:
@Ms2ger You should go ahead and review this.

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1207)

jdm commented:
```
▶ CRASH [expected TIMEOUT] /XMLHttpRequest/open-url-javascript-window-2.htm
  │ 
  │ ERROR:js::rust: Error at :0:0: parent.test is undefined
  │ 
  │ thread 'ScriptTask PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'assertion failed: `(left == right)` (left: `Some(Url { scheme: "about", scheme_data: NonRelative("blank"), query: None, fragment: None })`, right: `Some(Url { scheme: "javascript", scheme_data: NonRelative("parent.test.step(function() { var x = new XMLHttpRequest(); x.open(\'GET\', \'folder.txt\', false); x.send(null); parent.assert_equals(x.responseText, \'top\\n\'); parent.test.done() })"), query: None, fragment: None })`)', /Users/servo/buildbot/slave/mac-rel-wpt/build/components/script/dom/htmliframeelement.rs:228
  │ stack backtrace:
  │    1:        0x105834538 - sys::backtrace::tracing::imp::write::hca66179add43b2e5lLt
  │    2:        0x1058365ff - panicking::log_panic::_<closure>::closure.40842
  │    3:        0x1058360a2 - panicking::log_panic::h7a04edac7e1f2797HEx
  │    4:        0x105821eb6 - sys_common::unwind::begin_unwind_inner::h39a94087039ae0efnOs
  │    5:        0x10582229e - sys_common::unwind::begin_unwind_fmt::h7be0824d83765cd5tNs
  │    6:        0x104937946 - dom::htmliframeelement::_<impl>::iframe_load_event_steps::h2138545e1e82891dbwG
  │    7:        0x104a8fa79 - script_task::_<impl>::handle_msgs::_<closure>::closure.161623
  │    8:        0x104a779af - script_task::_<impl>::handle_msgs::ha2fd794f0901f8afkn6
  │    9:        0x104a3e487 - sys_common::unwind::try::try_fn::try_fn::h8464795135977377700
  │   10:        0x105833978 - __rust_try
  │   11:        0x105830a7e - sys_common::unwind::try::inner_try::h819d2da8e3dc4516VKs
  │   12:        0x104a3fdd1 - boxed::_<impl>::call_box::call_box::h7241207028857668738
  │   13:        0x10583588d - sys::thread::_<impl>::new::thread_start::h328957c252b95d3cOYw
  │   14:     0x7fff91449059 - _pthread_body
  │   15:     0x7fff91448fd6 - _pthread_start
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ CRASH [expected TIMEOUT] /XMLHttpRequest/open-url-javascript-window.htm
  │ 
  │ ERROR:js::rust: Error at :0:0: parent.request is not a function
  │ 
  │ thread 'ScriptTask PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'assertion failed: `(left == right)` (left: `Some(Url { scheme: "about", scheme_data: NonRelative("blank"), query: None, fragment: None })`, right: `Some(Url { scheme: "javascript", scheme_data: NonRelative("parent.request()"), query: None, fragment: None })`)', /Users/servo/buildbot/slave/mac-rel-wpt/build/components/script/dom/htmliframeelement.rs:228
  │ stack backtrace:
  │    1:        0x103c1d538 - sys::backtrace::tracing::imp::write::hca66179add43b2e5lLt
  │    2:        0x103c1f5ff - panicking::log_panic::_<closure>::closure.40842
  │    3:        0x103c1f0a2 - panicking::log_panic::h7a04edac7e1f2797HEx
  │    4:        0x103c0aeb6 - sys_common::unwind::begin_unwind_inner::h39a94087039ae0efnOs
  │    5:        0x103c0b29e - sys_common::unwind::begin_unwind_fmt::h7be0824d83765cd5tNs
  │    6:        0x102d20946 - dom::htmliframeelement::_<impl>::iframe_load_event_steps::h2138545e1e82891dbwG
  │    7:        0x102e78a79 - script_task::_<impl>::handle_msgs::_<closure>::closure.161623
  │    8:        0x102e609af - script_task::_<impl>::handle_msgs::ha2fd794f0901f8afkn6
  │    9:        0x102e27487 - sys_common::unwind::try::try_fn::try_fn::h8464795135977377700
  │   10:        0x103c1c978 - __rust_try
  │   11:        0x103c19a7e - sys_common::unwind::try::inner_try::h819d2da8e3dc4516VKs
  │   12:        0x102e28dd1 - boxed::_<impl>::call_box::call_box::h7241207028857668738
  │   13:        0x103c1e88d - sys::thread::_<impl>::new::thread_start::h328957c252b95d3cOYw
  │   14:     0x7fff91449059 - _pthread_body
  │   15:     0x7fff91448fd6 - _pthread_start
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ CRASH [expected OK] /dom/nodes/Document-URL.sub.html
  │ 
  │ thread 'ScriptTask PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'assertion failed: `(left == right)` (left: `Some(Url { scheme: "http", scheme_data: Relative(RelativeSchemeData { username: "", password: None, host: Domain("web-platform.test"), port: Some(8000), default_port: Some(80), path: ["common", "blank.html"] }), query: None, fragment: None })`, right: `Some(Url { scheme: "http", scheme_data: Relative(RelativeSchemeData { username: "", password: None, host: Domain("web-platform.test"), port: Some(8000), default_port: Some(80), path: ["common", "redirect.py"] }), query: Some("location=/common/blank.html"), fragment: None })`)', /Users/servo/buildbot/slave/mac-rel-wpt/build/components/script/dom/htmliframeelement.rs:228
  │ stack backtrace:
  │    1:        0x1102c7538 - sys::backtrace::tracing::imp::write::hca66179add43b2e5lLt
  │    2:        0x1102c95ff - panicking::log_panic::_<closure>::closure.40842
  │    3:        0x1102c90a2 - panicking::log_panic::h7a04edac7e1f2797HEx
  │    4:        0x1102b4eb6 - sys_common::unwind::begin_unwind_inner::h39a94087039ae0efnOs
  │    5:        0x1102b529e - sys_common::unwind::begin_unwind_fmt::h7be0824d83765cd5tNs
  │    6:        0x10f3ca946 - dom::htmliframeelement::_<impl>::iframe_load_event_steps::h2138545e1e82891dbwG
  │    7:        0x10f522a79 - script_task::_<impl>::handle_msgs::_<closure>::closure.161623
  │    8:        0x10f50a9af - script_task::_<impl>::handle_msgs::ha2fd794f0901f8afkn6
  │    9:        0x10f4d1487 - sys_common::unwind::try::try_fn::try_fn::h8464795135977377700
  │   10:        0x1102c6978 - __rust_try
  │   11:        0x1102c3a7e - sys_common::unwind::try::inner_try::h819d2da8e3dc4516VKs
  │   12:        0x10f4d2dd1 - boxed::_<impl>::call_box::call_box::h7241207028857668738
  │   13:        0x1102c888d - sys::thread::_<impl>::new::thread_start::h328957c252b95d3cOYw
  │   14:     0x7fff91449059 - _pthread_body
  │   15:     0x7fff91448fd6 - _pthread_start
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ CRASH [expected OK] /dom/nodes/Document-contentType/contentType/contenttype_javascripturi.html
  │ 
  │ thread 'ScriptTask PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'assertion failed: `(left == right)` (left: `Some(Url { scheme: "about", scheme_data: NonRelative("blank"), query: None, fragment: None })`, right: `Some(Url { scheme: "javascript", scheme_data: NonRelative("document.contentType"), query: None, fragment: None })`)', /Users/servo/buildbot/slave/mac-rel-wpt/build/components/script/dom/htmliframeelement.rs:228
  │ stack backtrace:
  │    1:        0x106984538 - sys::backtrace::tracing::imp::write::hca66179add43b2e5lLt
  │    2:        0x1069865ff - panicking::log_panic::_<closure>::closure.40842
  │    3:        0x1069860a2 - panicking::log_panic::h7a04edac7e1f2797HEx
  │    4:        0x106971eb6 - sys_common::unwind::begin_unwind_inner::h39a94087039ae0efnOs
  │    5:        0x10697229e - sys_common::unwind::begin_unwind_fmt::h7be0824d83765cd5tNs
  │    6:        0x105a87946 - dom::htmliframeelement::_<impl>::iframe_load_event_steps::h2138545e1e82891dbwG
  │    7:        0x105bdfa79 - script_task::_<impl>::handle_msgs::_<closure>::closure.161623
  │    8:        0x105bc79af - script_task::_<impl>::handle_msgs::ha2fd794f0901f8afkn6
  │    9:        0x105b8e487 - sys_common::unwind::try::try_fn::try_fn::h8464795135977377700
  │   10:        0x106983978 - __rust_try
  │   11:        0x106980a7e - sys_common::unwind::try::inner_try::h819d2da8e3dc4516VKs
  │   12:        0x105b8fdd1 - boxed::_<impl>::call_box::call_box::h7241207028857668738
  │   13:        0x10698588d - sys::thread::_<impl>::new::thread_start::h328957c252b95d3cOYw
  │   14:     0x7fff91449059 - _pthread_body
  │   15:     0x7fff91448fd6 - _pthread_start
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ CRASH [expected OK] /html/browsers/browsing-the-web/navigating-across-documents/javascript-url-query-fragment-components.html
  │ 
  │ thread 'ScriptTask PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'assertion failed: `(left == right)` (left: `Some(Url { scheme: "about", scheme_data: NonRelative("blank"), query: None, fragment: None })`, right: `Some(Url { scheme: "javascript", scheme_data: NonRelative("\"nope\" "), query: Some("%20%22yep%22%20:%20%22what%22;"), fragment: None })`)', /Users/servo/buildbot/slave/mac-rel-wpt/build/components/script/dom/htmliframeelement.rs:228
  │ stack backtrace:
  │    1:        0x109c88538 - sys::backtrace::tracing::imp::write::hca66179add43b2e5lLt
  │    2:        0x109c8a5ff - panicking::log_panic::_<closure>::closure.40842
  │    3:        0x109c8a0a2 - panicking::log_panic::h7a04edac7e1f2797HEx
  │    4:        0x109c75eb6 - sys_common::unwind::begin_unwind_inner::h39a94087039ae0efnOs
  │    5:        0x109c7629e - sys_common::unwind::begin_unwind_fmt::h7be0824d83765cd5tNs
  │    6:        0x108d8b946 - dom::htmliframeelement::_<impl>::iframe_load_event_steps::h2138545e1e82891dbwG
  │    7:        0x108ee3a79 - script_task::_<impl>::handle_msgs::_<closure>::closure.161623
  │    8:        0x108ecb9af - script_task::_<impl>::handle_msgs::ha2fd794f0901f8afkn6
  │    9:        0x108e92487 - sys_common::unwind::try::try_fn::try_fn::h8464795135977377700
  │   10:        0x109c87978 - __rust_try
  │   11:        0x109c84a7e - sys_common::unwind::try::inner_try::h819d2da8e3dc4516VKs
  │   12:        0x108e93dd1 - boxed::_<impl>::call_box::call_box::h7241207028857668738
  │   13:        0x109c8988d - sys::thread::_<impl>::new::thread_start::h328957c252b95d3cOYw
  │   14:     0x7fff91449059 - _pthread_body
  │   15:     0x7fff91448fd6 - _pthread_start
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ CRASH [expected TIMEOUT] /html/dom/documents/dom-tree-accessors/Document.currentScript.sub.html
  │ 
  │ ERROR:script::dom::htmlscriptelement: error loading script failed to lookup address information: nodename nor servname provided, or not known
  │ ERROR:js::rust: Error at http://web-platform.test:8000/html/dom/documents/dom-tree-accessors/Document.currentScript.sub.html:0:1: missing } in compound statement
  │ 
  │ ERROR:js::rust: Error at http://web-platform.test:8000/html/dom/documents/dom-tree-accessors/Document.currentScript.sub.html:0:0: undefinedfn is not defined
  │ 
  │ ERROR:js::rust: Error at http://web-platform.test:8000/html/dom/documents/dom-tree-accessors/Document.currentScript.sub.html:1:0: document.write is not a function
  │ 
  │ ERROR:js::rust: Error at :0:0: parent.verify is not a function
  │ 
  │ thread 'ScriptTask PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'assertion failed: `(left == right)` (left: `Some(Url { scheme: "about", scheme_data: NonRelative("blank"), query: None, fragment: None })`, right: `Some(Url { scheme: "javascript", scheme_data: NonRelative("parent.verify(\'iframe-src\'),parent.finish(\'iframe-src\')"), query: None, fragment: None })`)', /Users/servo/buildbot/slave/mac-rel-wpt/build/components/script/dom/htmliframeelement.rs:228
  │ stack backtrace:
  │    1:        0x10f0b0538 - sys::backtrace::tracing::imp::write::hca66179add43b2e5lLt
  │    2:        0x10f0b25ff - panicking::log_panic::_<closure>::closure.40842
  │    3:        0x10f0b20a2 - panicking::log_panic::h7a04edac7e1f2797HEx
  │    4:        0x10f09deb6 - sys_common::unwind::begin_unwind_inner::h39a94087039ae0efnOs
  │    5:        0x10f09e29e - sys_common::unwind::begin_unwind_fmt::h7be0824d83765cd5tNs
  │    6:        0x10e1b3946 - dom::htmliframeelement::_<impl>::iframe_load_event_steps::h2138545e1e82891dbwG
  │    7:        0x10e30ba79 - script_task::_<impl>::handle_msgs::_<closure>::closure.161623
  │    8:        0x10e2f39af - script_task::_<impl>::handle_msgs::ha2fd794f0901f8afkn6
  │    9:        0x10e2ba487 - sys_common::unwind::try::try_fn::try_fn::h8464795135977377700
  │   10:        0x10f0af978 - __rust_try
  │   11:        0x10f0aca7e - sys_common::unwind::try::inner_try::h819d2da8e3dc4516VKs
  │   12:        0x10e2bbdd1 - boxed::_<impl>::call_box::call_box::h7241207028857668738
  │   13:        0x10f0b188d - sys::thread::_<impl>::new::thread_start::h328957c252b95d3cOYw
  │   14:     0x7fff91449059 - _pthread_body
  │   15:     0x7fff91448fd6 - _pthread_start
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ CRASH [expected TIMEOUT] /html/semantics/embedded-content/the-iframe-element/iframe_javascript_url_01.htm
  │ 
  │ ERROR:js::rust: Error at about:blank:0:159: name is not defined
  │ 
  │ thread 'ScriptTask PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'assertion failed: `(left == right)` (left: `Some(Url { scheme: "about", scheme_data: NonRelative("blank"), query: None, fragment: None })`, right: `Some(Url { scheme: "javascript", scheme_data: NonRelative("\"<html><script>var sameDom = false; try{var cn = top.document.body.className;sameDom = true;}catch(e){}; parent.postMessage( {url: document.URL, name: name, sameDom: sameDom}, \'*\')</script><body><p>JS-generated document</p></body></<html>\";"), query: None, fragment: None })`)', /Users/servo/buildbot/slave/mac-rel-wpt/build/components/script/dom/htmliframeelement.rs:228
  │ stack backtrace:
  │ ERROR:js::rust: Error at about:blank:0:159: name is not defined
  │ 
  │    1:        0x10b977538 - sys::backtrace::tracing::imp::write::hca66179add43b2e5lLt
  │    2:        0x10b9795ff - panicking::log_panic::_<closure>::closure.40842
  │    3:        0x10b9790a2 - panicking::log_panic::h7a04edac7e1f2797HEx
  │    4:        0x10b964eb6 - sys_common::unwind::begin_unwind_inner::h39a94087039ae0efnOs
  │    5:        0x10b96529e - sys_common::unwind::begin_unwind_fmt::h7be0824d83765cd5tNs
  │    6:        0x10aa7a946 - dom::htmliframeelement::_<impl>::iframe_load_event_steps::h2138545e1e82891dbwG
  │ ERROR:js::rust: Error at about:blank:0:159: name is not defined
  │ 
  │ ERROR:js::rust: Error at about:blank:0:159: name is not defined
  │ 
  │    7:        0x10abd2a79 - script_task::_<impl>::handle_msgs::_<closure>::closure.161623
  │    8:        0x10abba9af - script_task::_<impl>::handle_msgs::ha2fd794f0901f8afkn6
  │    9:        0x10ab81487 - sys_common::unwind::try::try_fn::try_fn::h8464795135977377700
  │   10:        0x10b976978 - __rust_try
  │   11:        0x10b973a7e - sys_common::unwind::try::inner_try::h819d2da8e3dc4516VKs
  │   12:        0x10ab82dd1 - boxed::_<impl>::call_box::call_box::h7241207028857668738
  │   13:        0x10b97888d - sys::thread::_<impl>::new::thread_start::h328957c252b95d3cOYw
  │   14:     0x7fff91449059 - _pthread_body
  │   15:     0x7fff91448fd6 - _pthread_start
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ CRASH [expected OK] /old-tests/submission/Opera/script_scheduling/040.html
  │ 
  │ ERROR:js::rust: Error at :0:0: parent.log is not a function
  │ 
  │ thread 'ScriptTask PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'assertion failed: `(left == right)` (left: `Some(Url { scheme: "about", scheme_data: NonRelative("blank"), query: None, fragment: None })`, right: `Some(Url { scheme: "javascript", scheme_data: NonRelative("parent.log(\'JS URL\');\'<html><script>parent.log(\\\'frame script\\\')</script></html>\'"), query: None, fragment: None })`)', /Users/servo/buildbot/slave/mac-rel-wpt/build/components/script/dom/htmliframeelement.rs:228
  │ stack backtrace:
  │    1:        0x103760538 - sys::backtrace::tracing::imp::write::hca66179add43b2e5lLt
  │    2:        0x1037625ff - panicking::log_panic::_<closure>::closure.40842
  │    3:        0x1037620a2 - panicking::log_panic::h7a04edac7e1f2797HEx
  │    4:        0x10374deb6 - sys_common::unwind::begin_unwind_inner::h39a94087039ae0efnOs
  │    5:        0x10374e29e - sys_common::unwind::begin_unwind_fmt::h7be0824d83765cd5tNs
  │    6:        0x102863946 - dom::htmliframeelement::_<impl>::iframe_load_event_steps::h2138545e1e82891dbwG
  │    7:        0x1029bba79 - script_task::_<impl>::handle_msgs::_<closure>::closure.161623
  │    8:        0x1029a39af - script_task::_<impl>::handle_msgs::ha2fd794f0901f8afkn6
  │    9:        0x10296a487 - sys_common::unwind::try::try_fn::try_fn::h8464795135977377700
  │   10:        0x10375f978 - __rust_try
  │   11:        0x10375ca7e - sys_common::unwind::try::inner_try::h819d2da8e3dc4516VKs
  │   12:        0x10296bdd1 - boxed::_<impl>::call_box::call_box::h7241207028857668738
  │   13:        0x10376188d - sys::thread::_<impl>::new::thread_start::h328957c252b95d3cOYw
  │   14:     0x7fff91449059 - _pthread_body
  │   15:     0x7fff91448fd6 - _pthread_start
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ CRASH [expected OK] /old-tests/submission/Opera/script_scheduling/109.html
  │ 
  │ ERROR:js::rust: Error at :0:5: top.log is not a function
  │ 
  │ thread 'ScriptTask PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'assertion failed: `(left == right)` (left: `Some(Url { scheme: "about", scheme_data: NonRelative("blank"), query: None, fragment: None })`, right: `Some(Url { scheme: "javascript", scheme_data: NonRelative("void(top.log(\'JS URL\'));"), query: None, fragment: None })`)', /Users/servo/buildbot/slave/mac-rel-wpt/build/components/script/dom/htmliframeelement.rs:228
  │ stack backtrace:
  │    1:        0x110766538 - sys::backtrace::tracing::imp::write::hca66179add43b2e5lLt
  │    2:        0x1107685ff - panicking::log_panic::_<closure>::closure.40842
  │    3:        0x1107680a2 - panicking::log_panic::h7a04edac7e1f2797HEx
  │    4:        0x110753eb6 - sys_common::unwind::begin_unwind_inner::h39a94087039ae0efnOs
  │    5:        0x11075429e - sys_common::unwind::begin_unwind_fmt::h7be0824d83765cd5tNs
  │    6:        0x10f869946 - dom::htmliframeelement::_<impl>::iframe_load_event_steps::h2138545e1e82891dbwG
  │    7:        0x10f9c1a79 - script_task::_<impl>::handle_msgs::_<closure>::closure.161623
  │    8:        0x10f9a99af - script_task::_<impl>::handle_msgs::ha2fd794f0901f8afkn6
  │    9:        0x10f970487 - sys_common::unwind::try::try_fn::try_fn::h8464795135977377700
  │   10:        0x110765978 - __rust_try
  │   11:        0x110762a7e - sys_common::unwind::try::inner_try::h819d2da8e3dc4516VKs
  │   12:        0x10f971dd1 - boxed::_<impl>::call_box::call_box::h7241207028857668738
  │   13:        0x11076788d - sys::thread::_<impl>::new::thread_start::h328957c252b95d3cOYw
  │   14:     0x7fff91449059 - _pthread_body
  │   15:     0x7fff91448fd6 - _pthread_start
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ CRASH [expected OK] /_mozilla/mozilla/mozbrowser/redirect.html
  │ 
  │ thread 'ScriptTask PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'assertion failed: `(left == right)` (left: `Some(Url { scheme: "http", scheme_data: Relative(RelativeSchemeData { username: "", password: None, host: Domain("web-platform.test"), port: Some(8000), default_port: Some(80), path: ["_mozilla", "mozilla", "mozbrowser", "redirect_final.html"] }), query: None, fragment: None })`, right: `Some(Url { scheme: "http", scheme_data: Relative(RelativeSchemeData { username: "", password: None, host: Domain("web-platform.test"), port: Some(8000), default_port: Some(80), path: ["_mozilla", "mozilla", "mozbrowser", "redirect_init.html"] }), query: Some("pipe=status(302)|header(Location,redirect_final.html)"), fragment: None })`)', /Users/servo/buildbot/slave/mac-rel-wpt/build/components/script/dom/htmliframeelement.rs:228
  │ stack backtrace:
  │    1:        0x10e085538 - sys::backtrace::tracing::imp::write::hca66179add43b2e5lLt
  │    2:        0x10e0875ff - panicking::log_panic::_<closure>::closure.40842
  │    3:        0x10e0870a2 - panicking::log_panic::h7a04edac7e1f2797HEx
  │    4:        0x10e072eb6 - sys_common::unwind::begin_unwind_inner::h39a94087039ae0efnOs
  │    5:        0x10e07329e - sys_common::unwind::begin_unwind_fmt::h7be0824d83765cd5tNs
  │    6:        0x10d188946 - dom::htmliframeelement::_<impl>::iframe_load_event_steps::h2138545e1e82891dbwG
  │    7:        0x10d2e0a79 - script_task::_<impl>::handle_msgs::_<closure>::closure.161623
  │    8:        0x10d2c89af - script_task::_<impl>::handle_msgs::ha2fd794f0901f8afkn6
  │    9:        0x10d28f487 - sys_common::unwind::try::try_fn::try_fn::h8464795135977377700
  │   10:        0x10e084978 - __rust_try
  │   11:        0x10e081a7e - sys_common::unwind::try::inner_try::h819d2da8e3dc4516VKs
  │   12:        0x10d290dd1 - boxed::_<impl>::call_box::call_box::h7241207028857668738
  │   13:        0x10e08688d - sys::thread::_<impl>::new::thread_start::h328957c252b95d3cOYw
  │   14:     0x7fff91449059 - _pthread_body
  │   15:     0x7fff91448fd6 - _pthread_start
  └ Pipeline failed in hard-fail mode.  Crashing!
```

jdm commented:
Fixed all the remaining test failures in the latest push.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #9214) made this pull request unmergeable. Please resolve the merge conflicts.

frewsxcv commented:
`./components/script/script_thread.rs:1129: missing space before }`

bors-servo commented:
:umbrella: The latest upstream changes (presumably #9263) made this pull request unmergeable. Please resolve the merge conflicts.

Ms2ger committed:
Rewrite and clarify base_multiple.html.

Ms2ger committed:
Ensure that base_multiple.html fails rather than times out in the typical failure case.

Ms2ger commented:
Finally made it through this PR; looks great!

Please squash the first two commits, and revert random style changes to files that are now no longer changed otherwise.

```diff
+    /// Remove this load from the associated document's list of blocking loads.
+    pub fn terminate(&mut self) {
```

I believe this should take `self`. Destructuring the value would ensure the destructor won't run, so you could just panic there.

```diff
+    /// Return the url associated with this load.
+    pub fn url(&self) -> Option<&Url> {
```

Is this used?

```diff
+    pub fn add_blocking_load(&mut self, load: LoadType) {
```

Docs.

```diff
+    pub fn add_blocking_load(&self, load: LoadType) {
```

Docs.

```diff
+        let document = document.r();
```

Drop this?

```diff
+        let mut load_blocker = self.load_blocker.borrow_mut();
+        if let Some(ref mut load_blocker) = load_blocker.as_mut() {
+            // Any oustanding load is finished from the point of view of the blocked
+            // document; the new navigation will continue blocking it.
+            load_blocker.terminate();
```

This causes a panic with `--pref dom.mozbrowser.enabled` and:
```html
.
<script>
var iframe = document.createElement("iframe");
document.body.appendChild(iframe);
iframe.reload();
iframe.reload();
</script>
```

I think you want `load_blocker.take()`.

```diff
+        //TODO: Deal with the case where an iframe is being reloaded so url is None.
```

File an issue.

```diff
--- /dev/null
+++ b/tests/wpt/metadata/html/browsers/history/the-location-interface/location_reload.html.ini
@@ -0,0 +1,3 @@
+[location_reload.html]
+  type: testharness
+  expected: CRASH
```

What's causing this crash?

```diff
--- a/tests/wpt/metadata/html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute.html.ini
+++ b/tests/wpt/metadata/html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute.html.ini
@@ -1,3 +1,1827 @@
 [parse-a-sizes-attribute.html]
   type: testharness
-  expected: TIMEOUT
+  disabled: unsupported feature and prone to intermittent incorrect results
```

File an issue, or mention in a related issue, or something.

jdm commented:
Reviewed 1 of 1 files at r6, 1 of 2 files at r7.
Review status: 1 of 19 files reviewed at latest revision, 11 unresolved discussions.

---

*[tests/wpt/web-platform-tests/html/semantics/document-metadata/the-base-element/base_multiple.html, line 8 \[r6\]](https://reviewable.io:443/reviews/servo/servo/6677#-K9m6SXeTLvxgQgRdmfx:-K9m6SXeTLvxgQgRdmfy:-800411758) ([raw file](https://github.com/servo/servo/blob/346a754c60d8580363e7fbe7e1bf4a9a70743d23/tests/wpt/web-platform-tests/html/semantics/document-metadata/the-base-element/base_multiple.html#L8)):*
Let's remove this.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6677)*
<!-- Sent from Reviewable.io -->

jdm commented:
The difficulty with making `terminate` take self by value is this:
```
/Users/jdm/src/servo/components/script/dom/htmliframeelement.rs:109:21: 109:37 error: Expression of type document_loader::LoadBlocker must be rooted, #[deny(unrooted_must_root)] on by default
/Users/jdm/src/servo/components/script/dom/htmliframeelement.rs:109         if let Some(mut load_blocker) = load_blocker.take() {
                                                                                        ^~~~~~~~~~~~~~~~
/Users/jdm/src/servo/components/script/dom/htmliframeelement.rs:109:21: 109:37 warning: variable does not need to be mutable, #[warn(unused_mut)] on by default
/Users/jdm/src/servo/components/script/dom/htmliframeelement.rs:109         if let Some(mut load_blocker) = load_blocker.take() {
                                                                                        ^~~~~~~~~~~~~~~~
/Users/jdm/src/servo/components/script/dom/htmliframeelement.rs:242:21: 242:32 error: Expression of type document_loader::LoadBlocker must be rooted, #[deny(unrooted_must_root)] on by default
/Users/jdm/src/servo/components/script/dom/htmliframeelement.rs:242         if let Some(mut blocker) = blocker.take() {
                                                                                        ^~~~~~~~~~~
/Users/jdm/src/servo/components/script/dom/htmliframeelement.rs:242:21: 242:32 warning: variable does not need to be mutable, #[warn(unused_mut)] on by default
/Users/jdm/src/servo/components/script/dom/htmliframeelement.rs:242         if let Some(mut blocker) = blocker.take() {
                                                                                        ^~~~~~~~~~~
/Users/jdm/src/servo/components/script/dom/htmliframeelement.rs:527:21: 527:32 error: Expression of type document_loader::LoadBlocker must be rooted, #[deny(unrooted_must_root)] on by default
/Users/jdm/src/servo/components/script/dom/htmliframeelement.rs:527         if let Some(mut blocker) = blocker.take() {
                                                                                        ^~~~~~~~~~~
/Users/jdm/src/servo/components/script/dom/htmliframeelement.rs:527:21: 527:32 warning: variable does not need to be mutable, #[warn(unused_mut)] on by default
/Users/jdm/src/servo/components/script/dom/htmliframeelement.rs:527         if let Some(mut blocker) = blocker.take() {
                                                                                        ^~~~~~~~~~~
```
I don't see a way around this, unfortunately.

jdm committed:
Make location_reload.html not trigger resource exhaustion.

jdm committed:
Improve documentation for new methods.

jdm committed:
Add a test for panicking when reloading an iframe twice.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit f7868a5 with merge 42b92ce...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1636)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #9571) made this pull request unmergeable. Please resolve the merge conflicts.

Ms2ger commented:
Maybe we could make `terminate` take `&mut Option<LoadBlocker>`?

```diff
+  var loaded = false;
+  iframe.onload = t.step_func(function() {
+    if (!loaded) {
+      loaded = true;
+      iframe.reload();
+      iframe.reload();
+    } else {
+      t.done();
+    }
+  });
```

Should this wait for the third load event? Or we we just dispatch two?

Please also address my other comments.

jdm commented:
@Ms2ger Which comments did I not address?

jdm commented:
With regards to the `fn url` that isn't used, it's convenient for debugging so I'd rather not remove it.

jdm commented:
Filed #9592 and #9560.

jdm commented:
With regards to the test, we only get two load events.

jdm commented:
@bors-servo: r=Ms2ger

bors-servo commented:
:pushpin: Commit c5fe8f7 has been approved by `Ms2ger`

<!-- @bors-servo r=Ms2ger c5fe8f74010a872e1b0a2f5d7f956c679fc08855 -->

bors-servo commented:
:hourglass: Testing commit c5fe8f7 with merge a31f31e...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/313)

jdm commented:
@bors-servo: retry
* git issue

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/1608), [gonk](http://build.servo.org/builders/gonk/builds/1641), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1614), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1835), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1666), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1698) are reusable. Rebuilding only [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/313)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1608), [gonk](http://build.servo.org/builders/gonk/builds/1641), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1614), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1835), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/314), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1666), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1698)
